[DCP][OSS] Rank local checkpointing in DCP without collectives

saumishr opened the pull request:
Summary:
DCP metadata collectives become prohibitively expensive as the job scale grows. This PR introduces rank-local checkpointing which basically saves and loads the checkpoint without any collective. The trade off for now is the dedupe and re-sharding. Support for these would be introduced soon.

Differential Revision: D70112642




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @LucasLLC @pradeepfn @kwen2501 @c-p-i-o @MeetVadakkanchery @mhorowitz @ekr0

meetv18 reviewed (commented):
Overall save path LGTM, but I have some concerns on load. Happy to approve after resolved. Thanks!

saumishr commented on torch/distributed/checkpoint/default_planner.py:
```diff
@@ -654,7 +654,7 @@ def _validate_global_plan(global_plan: list[SavePlan], metadata: Metadata) -> bo
         # Check whether combined chunk cover the whole tensor
         tensor_volume = reduce(operator.mul, value.size, 1)
-        if chunks_volume != tensor_volume:
+        if len(global_plan) > 1 and chunks_volume != tensor_volume:
```
Tensor volume check makes sense only in the global context. When every rank is doing its own planning, this check doesn't have much value. I plan to refactor the plan validation into local and global validation and then it would become cleaner.

saumishr committed:
[DCP][OSS] Rank local checkpointing in DCP without collectives (#147758)

Summary:
X-link: https://github.com/pytorch/tnt/pull/991



Context:
DCP metadata collectives become prohibitively expensive as the job scale grows. This PR introduces rank-local checkpointing which basically saves and loads the checkpoint without any collective. The trade off for now is the dedupe and re-sharding. Support for these would be introduced soon.

Test Plan:
E2E UTs

Save and load test with internal DCP components: https://www.internalfb.com/mlhub/pipelines/runs/mast/torchx-textray-pretrain_mlm-lv5d7qcfmnqzkd

Save and load test with OSS DCP components: https://www.internalfb.com/mlhub/pipelines/runs/mast/torchx-textray-pretrain_mlm-z1vz46vkkgtcld
https://www.internalfb.com/mlhub/pipelines/runs/mast/torchx-textray-pretrain_mlm-njvvbn07rv5ckd

Reviewed By: meetv18

Differential Revision: D70112642
